pkg/apis/storage/validation/validation.go

```diff
@@ -383,14 +383,16 @@ func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets
 		allErrs = append(allErrs, field.Duplicate(fldPath.Child("name"), driver.Name))
 	}
 	driverNamesInSpecs.Insert(driver.Name)
-	topoKeys := make(sets.String)
+	topoKeys := sets.New[string]()
 	for _, key := range driver.TopologyKeys {
 		if len(key) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath, key))
+			continue
 		}
 
 		if topoKeys.Has(key) {
 			allErrs = append(allErrs, field.Duplicate(fldPath, key))
+			continue
```

Comment:
What value does it add? The only effect this would have is that we don't insert a key more than once to `topoKeys` which should not matter because its only purpose is to track duplicates here.

What bug are we trying to solve exactly? The [tracking issue](https://github.com/kubernetes/kubernetes/issues/117027) does not explain it well, but sounds like we want to avoid doing the validation twice - which I don't think will happen with the original code. Do we have an evidence/test to prove this maybe?

Reply:
I think the problem we trying to solve here is more like a cleanup, to avoid doing validation twice. and the tracking issue didn't provide more info. and I just use my judgement and found out it seems reasonable for me to add this.

Reply:
I think in this part will cause valid twice, while validaing CSINode Update. this func will finally call to the ``validateCSINodeDriver()``
https://github.com/kubernetes/kubernetes/blob/2f1db33dd54cbdba52c21c75a53c5387f1cd14ab/pkg/apis/storage/validation/validation.go#L300

Reply:
@charles-chenzz  I've checked the current code with unit test, adding a few prints to see how many times `validateCSINodeDriver()` executes. This verifies we correctly validate once for every driver:


```$ go test -v -run TestCSINodeUpdateValidation
=== RUN   TestCSINodeUpdateValidation
Starting success test case for drivers: [io.kubernetes.storage.csi.driver-1 io.kubernetes.storage.csi.driver-2]
VALIDATING: validateCSINodeDriver

VALIDATING: validateCSINodeDriver

Starting success test case for drivers: [io.kubernetes.storage.csi.driver-1]
VALIDATING: validateCSINodeDriver

Starting success test case for drivers: [io.kubernetes.storage.csi.driver-1 io.kubernetes.storage.csi.driver-2 io.kubernetes.storage.csi.driver-3]
VALIDATING: validateCSINodeDriver

VALIDATING: validateCSINodeDriver

VALIDATING: validateCSINodeDriver

```

If you still believe there is a bug please write a unit test that exposes this bug - it should be failing without the proposed patch, and passing with the patch applied.

Reply:
I tried to write a unit test that expose it but couldn't. so what I think is wrong tbh. how about we remove continue here and only replace the generic set? can we merge that?

Reply:
@charles-chenzz that sounds good to me, thanks!